openpype/modules/kitsu/plugins/publish/collect_kitsu_entities.py

```diff
@@ -13,61 +11,61 @@ class CollectKitsuEntities(pyblish.api.ContextPlugin):
 
     def process(self, context):
 
-        asset_data = context.data["assetEntity"]["data"]
-        zou_asset_data = asset_data.get("zou")
-        if not zou_asset_data:
-            raise AssertionError("Zou asset data not found in OpenPype!")
-        self.log.debug("Collected zou asset data: {}".format(zou_asset_data))
-
-        zou_task_data = asset_data["tasks"][os.environ["AVALON_TASK"]].get(
-            "zou"
-        )
-        if not zou_task_data:
-            self.log.warning("Zou task data not found in OpenPype!")
-        self.log.debug("Collected zou task data: {}".format(zou_task_data))
+        kitsu_project = None
 
-        kitsu_project = gazu.project.get_project(zou_asset_data["project_id"])
-        if not kitsu_project:
-            raise AssertionError("Project not found in kitsu!")
-        context.data["kitsu_project"] = kitsu_project
-        self.log.debug("Collect kitsu project: {}".format(kitsu_project))
+        kitsu_entities_by_id = {}
+        for instance in context:
+            asset_doc = instance.data.get("assetEntity")
+            task_name = instance.data.get("task")
+            if not asset_doc:
+                continue
 
-        entity_type = zou_asset_data["type"]
-        if entity_type == "Shot":
-            kitsu_entity = gazu.shot.get_shot(zou_asset_data["id"])
-        else:
-            kitsu_entity = gazu.asset.get_asset(zou_asset_data["id"])
+            zou_asset_data = asset_doc["data"].get("zou")
+            if not zou_asset_data:
+                raise ValueError("Zou asset data not found in OpenPype!")
 
-        if not kitsu_entity:
-            raise AssertionError("{} not found in kitsu!".format(entity_type))
+            if kitsu_project is None:
+                kitsu_project = gazu.project.get_project(
```

Comment:
I could do that but if, for some reason, an instance Kitsu project isn't the same as the rest this won't be catched.
I can't really see this happening but who knows? Should I still move it outside the loop?

Reply:
I think project name matching is the very less we could expect from the users. Mainly because most of the time, the sync is from Kitsu to OP, therefore no error is possible. If we'd really want to ensure it, we should create the same link between entities with IDs we do for the rest. But the current proposal looks unrelevant to me. I'd agree to move it outside the loop.